### PR TITLE
gnrc/rpl: document implementation status

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

This copies the information from #3662 to the documentation.
Makes it more discoverable and let's people manage their expectations.

### Testing procedure

Read the documentation.
Did I miss to mention everything?

I noticed routes downward in the tree don't work for all nodes. Should this be documented?

### Issues/PRs references

 #3662

### CODE REVIEW

**Comment by cgundogan** [comment]
>I noticed routes downward in the tree don't work for all nodes. Should this be documented?

With how many nodes did you try this?  Usually, this problem occurs if the FIB is smaller than the number of nodes in the network.

**Comment by benpicco** [comment]
> With how many nodes did you try this? Usually, this problem occurs if the FIB is smaller than the number of nodes in the network.

I tried with 50 nodes on the IoT Lab and set `GNRC_IPV6_FIB_TABLE_SIZE=50`.
Is there another parameter that needs to be tweaked?

**Comment by cgundogan** [comment]
>I tried with 50 nodes on the IoT Lab testbed and set GNRC_IPV6_FIB_TABLE_SIZE=50.
Is there another parameter that needs to be tweaked?

this parameter is outdated and should be removed from the code base .. (@miri64 do you think otherwise?).

The parameter you need is here:
https://github.com/RIOT-OS/RIOT/blob/master/sys/include/net/gnrc/ipv6/nib/conf.h#L271

**Comment by cgundogan** [comment]
Maybe also add the NIB parameter do the doc to reduce confusion?

**Comment by benpicco** [comment]
> The parameter you need is here:
> https://github.com/RIOT-OS/RIOT/blob/master/sys/include/net/gnrc/ipv6/nib/conf.h#L271

That didn't help.
I also tried with

```
CFLAGS += -DGNRC_IPV6_FIB_TABLE_SIZE=50
CFLAGS += -DCONFIG_GNRC_IPV6_NIB_OFFL_NUMOF=50
CFLAGS += -DCONFIG_GNRC_IPV6_NIB_L2ADDR_MAX_LEN=50
```

just to be sure.

**Comment by miri64** [comment]
> That didn't help.
> I also tried with
> 
> ```
> CFLAGS += -DGNRC_IPV6_FIB_TABLE_SIZE=50
> CFLAGS += -DCONFIG_GNRC_IPV6_NIB_OFFL_NUMOF=50
> ```
> 
> just to be sure.

You also need to make sure there are enough entries where those off-link entries are pointing to: https://github.com/RIOT-OS/RIOT/blob/master/sys/include/net/gnrc/ipv6/nib/conf.h#L260-L262. The `fib` module is not used, so even if you try to make sure, the `GNRC_IPV6_FIB_TABLE_SIZE` configuration option is ignored ;-).

> this parameter is outdated and should be removed from the code base .. (@miri64 do you think otherwise?).

The `fib` module _is_ still part of the code base, but we should strip this configuration option from the examples IMHO.

**Comment by miri64** [comment]
@benpicco what tool did you use the visualize the DODAG :star_struck:?

**Comment by miri64** [comment]
The help of your tool should explain what the white lines are, I guess. But my bet is on alternate parents.

**Comment by benpicco** [comment]
> You also need to make sure there are enough entries where those off-link entries are pointing to: https://github.com/RIOT-OS/RIOT/blob/master/sys/include/net/gnrc/ipv6/nib/conf.h#L260-L262. 

Thank you, I'll try again with `CFLAGS += -DCONFIG_GNRC_IPV6_NIB_NUMOF=50` :smile: 
Though the NIB would only be about *direct* neighbors. But guess everyone's a neighbor when you share a WPAN? 

> @benpicco what tool did you use the visualize the DODAG ￼?

That's [Foren6](https://cetic.github.io/foren6/). It's described in [this tutorial](https://github.com/iot-lab/iot-lab/wiki/Tutorial-about-RIOT-Operating-System-for-IOT-LAB-M3-nodes).

I used 

<details><summary>this script</summary>

```bash
#!/bin/bash

NODES="86 93 118 119 126"
REMOT_PORT=20000
LOCAL_PORT=40000

TTY_N=0

for n in $NODES; do
    ((TTY_N=TTY_N+1))
    ((LOCAL_PORT=LOCAL_PORT+1))
    screen -d -m ssh grenoble -L $LOCAL_PORT:m3-$n:$REMOT_PORT
    sleep 1
    screen -d -m socat TCP4:127.0.0.1:$LOCAL_PORT pty,link=/tmp/mytty$TTY_N,raw
done
```

</details>

to set up the forwarding to the sniffer nodes.

> The help of your tool should explain what the white lines are, I guess. But my bet is on alternate parents.

It's [a bit short](https://cetic.github.io/foren6/guide.html), but that makes sense.
Occasionally I get "Dodag config mismatch errors" on some nodes, not sure what those are about.

**Comment by benpicco** [comment]
> Thank you, I'll try again with `CFLAGS += -DCONFIG_GNRC_IPV6_NIB_NUMOF=50`

This did the trick!
Now I have all nodes in `nib route` and I can reach them all from the root.

So just to clarify (and update the doc while I'm at it): For this to work we need to set *both* `CONFIG_GNRC_IPV6_NIB_NUMOF` *and* `CONFIG_GNRC_IPV6_NIB_OFFL_NUMOF` to the number of nodes?

File: sys/include/net/gnrc/rpl.h [hunk 1]
```diff
@@ -82,6 +82,17 @@
  *   CFLAGS += -DCONFIG_GNRC_RPL_WITHOUT_VALIDATION
  *   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  *
+ * TODO
+ * ------
+ *
+ * The GNRC RPL implementation only implements storing mode with OF0 ([RFC6552](https://tools.ietf.org/html/rfc6552)).
+ * For interoparability with other RPL interoperability, open task include:
+ *
+ * - IPv6 Hop-by-hop RPL option
```
Comment (miri64) [inline sys/include/net/gnrc/rpl.h:91]: Currently in (slow) progress :-).

**Comment by miri64** [comment]
>  For this to work we need to set _both_ `CONFIG_GNRC_IPV6_NIB_NUMOF` _and_ `CONFIG_GNRC_IPV6_NIB_OFFL_NUMOF` to the number of nodes?

What is meant with "this"?

**Comment by benpicco** [comment]
> What is meant with "this"?

Downward routes to all nodes.

**Comment by miri64** [comment]
> > What is meant with "this"?
> 
> Downward routes to all (50) nodes.

In the specific yes, in general, no.

The NIB is in its core consistent of 2 data-structures: off-link entries (capped of by `CONFIG_GNRC_IPV6_NIB_OFFL_NUMOF`) and on-link entries (capped of by `CONFIG_GNRC_IPV6_NIB_NUMOF`). Off-link entries can be part of the FIB (but do not have to be). If they are, their next hop component is made up of an on-link entry. So very simply said: the off-link entry is prefix + prefix length and the on-link entry the next hop address + interface.

So yes, if you have 50 128-bit prefixes pointing to 50 different next hop addresses, you need to configure both to 50. You can however also have 50 routes with many routes pointing to the same downstream nodes. Then you may configure `NIB_NUMOF < NIB_OFFL_NUMOF`, or many non-RPL leaves at a router which are handled by ND, then you may configure `NIB_OFFL_NUMOF < NIB_NUMOF`.

**Comment by benpicco** [comment]
> You can however also have 50 routes with many routes pointing to the same downstream nodes. Then you may configure `NIB_NUMOF < NIB_OFFL_NUMOF`, or many non-RPL leaves at a router which are handled by ND, then you may configure `NIB_OFFL_NUMOF < NIB_NUMOF`.

I'm not sure how to translate those constraints into a general documentation.

If we have n nodes in the network and want to reach each node from the RPL root and don't know the topology of the network, we would have `CONFIG_GNRC_IPV6_NIB_NUMOF == CONFIG_GNRC_IPV6_NIB_OFFL_NUMOF == n`, right?

File: sys/include/net/gnrc/rpl.h [hunk 2]
```diff
@@ -82,6 +82,17 @@
  *   CFLAGS += -DCONFIG_GNRC_RPL_WITHOUT_VALIDATION
  *   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  *
+ * TODO
+ * ------
+ *
+ * The GNRC RPL implementation only implements storing mode with OF0 ([RFC6552](https://tools.ietf.org/html/rfc6552)).
+ * For interoparability with other RPL interoperability, open task include:
```
Comment (benpicco) [inline sys/include/net/gnrc/rpl.h:89]:
```suggestion
 * For interoperability with other RPL implementations, open task include:
```

**Comment by miri64** [comment]
> If we have n nodes in the network and want to reach each node from the RPL root and don't know the topology of the network, we would have `CONFIG_GNRC_IPV6_NIB_NUMOF == CONFIG_GNRC_IPV6_NIB_OFFL_NUMOF == n`, right?

Exactly :-)

File: sys/include/net/gnrc/rpl.h [hunk 3]
```diff
@@ -82,6 +82,17 @@
  *   CFLAGS += -DCONFIG_GNRC_RPL_WITHOUT_VALIDATION
  *   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  *
+ * TODO
+ * ------
+ *
+ * The GNRC RPL implementation only implements storing mode with OF0 ([RFC6552](https://tools.ietf.org/html/rfc6552)).
```
Comment (miri64) [inline sys/include/net/gnrc/rpl.h:88]: Maybe also add the sentence

```suggestion
 * The GNRC RPL implementation only implements storing mode with OF0 ([RFC6552](https://tools.ietf.org/html/rfc6552)).
 * The RPL routing header is parsed by the nodes, but anything else for non-storing mode is missing.
```

Maybe the more appropriate place would be with `Non-storing mode` below. Not sure...

**Comment by benpicco** [comment]
> So yes, if you have 50 128-bit prefixes pointing to 50 different next hop addresses, you need to configure both to 50. 

Just to clarify: The fact that all nodes have the same 64 bit prefix is orthogonal to this?

> You can however also have 50 routes with many routes pointing to the same downstream nodes. Then you may configure `NIB_NUMOF < NIB_OFFL_NUMOF`

Does that also include alternate parents?

**Comment by miri64** [comment]
> > 
> 
> Just to clarify: The fact that all nodes have the same 64 bit prefix is orthogonal to this?

Yes, the down-stream nodes in storing mode are 128-bit prefixes which contain the 64-bit prefix.

**Comment by miri64** [comment]
> > You can however also have 50 routes with many routes pointing to the same downstream nodes. Then you may configure `NIB_NUMOF < NIB_OFFL_NUMOF`
> 
> Does that also include alternate parents?

Yepp, may be.

**Comment by benpicco** [comment]
> > > You can however also have 50 routes with many routes pointing to the same downstream nodes. Then you may configure `NIB_NUMOF < NIB_OFFL_NUMOF`
> > 
> > 
> > Does that also include alternate parents?
> 
> Yepp, may be.

Hmm so this is actually the common case with RPL.

So `3 * CONFIG_GNRC_IPV6_NIB_NUMOF` == `CONFIG_GNRC_IPV6_NIB_OFFL_NUMOF` might be the better recommendation here.

**Comment by miri64** [comment]
> So `3 * CONFIG_GNRC_IPV6_NIB_NUMOF` == `CONFIG_GNRC_IPV6_NIB_OFFL_NUMOF` might be the better recommendation here.

Huh? No, if you have 50 nodes in the network you need at most `CONFIG_GNRC_IPV6_NIB_NUMOF = 50`. The number of off-link entries might need to be larger however.

**Comment by miri64** [comment]
Off-link entries have a pointer to on-link entries, so on-link entries can be re-used for different off-link entries.

**Comment by miri64** [comment]
And for alternate parents its actually [CONFIG_GNRC_IPV6_NIB_DEFAULT_ROUTER_NUMOF](https://doc.riot-os.org/group__net__gnrc__ipv6__nib__conf.html#ga67f93b381ce2aad915fa59b99b1b6da5) you need to increase, since they are modeled as a default route (basically another class in the NIB, that are off-link entries that don't have a prefix).

**Comment by benpicco** [comment]
Ok, so we have

```
CFLAGS += -DCONFIG_GNRC_IPV6_NIB_NUMOF=n
CFLAGS += -DCONFIG_GNRC_IPV6_NIB_OFFL_NUMOF=n * CONFIG_GNRC_IPV6_NIB_DEFAULT_ROUTER_NUMOF
```

**Comment by miri64** [comment]
> Ok, so we have
> 
> ```
> CFLAGS += -DCONFIG_GNRC_IPV6_NIB_NUMOF=n
> CFLAGS += -DCONFIG_GNRC_IPV6_NIB_OFFL_NUMOF=n * CONFIG_GNRC_IPV6_NIB_DEFAULT_ROUTER_NUMOF
> ```

Mhh no. Off-link entries and default routers are distinct arrays. Their maximum number does not depend on each other.

**Comment by miri64** [comment]
You only need to increase `CONFIG_GNRC_IPV6_NIB_DEFAULT_ROUTER_NUMOF` if you want to store more than one default route.

**Comment by miri64** [comment]
Off-link entries are only used up for non-default routes.

**Comment by benpicco** [comment]
> Off-link entries are only used up for non-default routes.

Ok, so for the rpl root that means one route for each node, right? If nodes can have alternative routes (`CONFIG_GNRC_IPV6_NIB_DEFAULT_ROUTER_NUMOF` > 1) we would have to store > n routes here.

**Comment by miri64** [comment]
> If nodes can have alternative routes (`CONFIG_GNRC_IPV6_NIB_DEFAULT_ROUTER_NUMOF` > 1) we would have to store > n routes here.

That _could_ be done, but I don't think it is needed. Usually you want 1 maybe 2 alternate parents in case the current parent is offline (and even then the DODAG could just repair itself). Just to have the potential for _all_ nodes to be the parent would be overkill IMHO.

**Comment by miri64** [comment]
To cut this discussion short: I think your initial proposal of `CONFIG_GNRC_IPV6_NIB_NUMOF=n` and `CONFIG_GNRC_IPV6_NIB_OFFL_NUMOF=n` is most sensible and maybe mention that `CONFIG_GNRC_IPV6_NIB_DEFAULT_ROUTER_NUMOF` cuts of the number of parents a node can have.

**Comment by benpicco** [comment]
I added a note.

File: sys/include/net/gnrc/rpl.h [hunk 4]
```diff
@@ -82,6 +82,42 @@
  *   CFLAGS += -DCONFIG_GNRC_RPL_WITHOUT_VALIDATION
  *   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  *
+ * - This RPL implementation currently only supports storing mode.
+ *   That means, in order to have downwards routes to all nodes the
+ *   Neighbor Information Base must be big enough to store information for
```
Comment (miri64) [inline sys/include/net/gnrc/rpl.h:87]: What is a Neighbor Information Base? ;-)

File: sys/include/net/gnrc/rpl.h [hunk 5]
```diff
@@ -82,6 +82,42 @@
...
+ *   each node.
+ *   To store the actual routes, the number of off-link entries in NIB
```
Comment (miri64) [inline sys/include/net/gnrc/rpl.h:89]: What are off-link entries?

**Review by miri64 — COMMENTED** [review]
You use terms that are basically proprietary to RIOT. People not that familiar with our IPv6 internals might be confused by that.

File: sys/include/net/gnrc/rpl.h (hunk 4 above)
Comment (benpicco) [reply sys/include/net/gnrc/rpl.h:87]: You know better than me ;)

What do you suggest to put here? After all, those tunables are very much RIOT specific.

I would very much welcome a glossary in the documentation. I've scratched my head more than once what those TLAs in the network stack all *mean*.

Comment (miri64) [reply sys/include/net/gnrc/rpl.h:87]: IMHO it is enough to add a link here.

```
 *   That means, in order to have downwards routes to all nodes the
 *   storage space within `gnrc_ipv6`'s [Neighbor Information Base](@ref net_gnrc_ipv6_nib)
 *   must be big enough to store information for each node.
```

File: sys/include/net/gnrc/rpl.h (hunk 5 above)
Comment (miri64) [reply sys/include/net/gnrc/rpl.h:89]: Too much implementation detail on the NIB IMHO. This sentence can be dropped. That two configuration parameters need to be changed should be clear from the rest. The interested reader can then look at the NIB itself, to find out why that is ;-).

File: sys/include/net/gnrc/rpl.h (hunk 4 above)
Comment (miri64) [reply sys/include/net/gnrc/rpl.h:87]: The term NIB is an invention of mine, as it is not just a neighbor cache (that's why the N stands for "Neighbor") but also more than a FIB. The comparison I always draw is Linux's `iproute2` which is also handling the ARP tables and the IPv6 neighbor cache on Linux.

File: sys/include/net/gnrc/rpl.h [hunk 6]
```diff
@@ -82,6 +82,40 @@
  *   CFLAGS += -DCONFIG_GNRC_RPL_WITHOUT_VALIDATION
  *   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  *
+ * - This RPL implementation currently only supports storing mode.
+ *   That means, in order to have downwards routes to all nodes the
+ *   storage space within [`gnrc_ipv6`'s Neighbor Information Base](@ref net_gnrc_ipv6_nib)
```
Comment (miri64) [inline sys/include/net/gnrc/rpl.h:87]: Markdown code markers (`) in a link [don't seem to work](https://3826-7370241-gh.circle-artifacts.com/0/doc/group__net__gnrc__rpl.html) :-( Then let's just use Doxygen markup:

```suggestion
 *   storage space within [@c gnrc_ipv6's Neighbor Information Base](@ref net_gnrc_ipv6_nib)
```

**Review by miri64 — CHANGES_REQUESTED** [review]

File: sys/include/net/gnrc/rpl.h [hunk 7]
```diff
@@ -82,6 +82,40 @@
...
+ *
+ * TODO
+ * ------
+ *
+ * The GNRC RPL implementation only implements storing mode with OF0 ([RFC6552](https://tools.ietf.org/html/rfc6552)).
+ * The RPL routing header is parsed by the nodes, but anything else for non-storing mode is missing.
+ * For interoperability with other RPL implementations, open task include:
+ *
+ * - IPv6 Hop-by-hop RPL option
+ * - Metric based routing ([RFC6551](https://tools.ietf.org/html/rfc6551) and [RFC6719](https://tools.ietf.org/html/rfc6719))
```
Comment (miri64) [inline sys/include/net/gnrc/rpl.h:115]: Line-length

File: sys/include/net/gnrc/rpl.h [hunk 8]
```diff
@@ -82,6 +82,40 @@
...
+ * - Non-Storing mode
+ * - DAG-Metric Container ([RFC6550#6.7.4](https://tools.ietf.org/html/rfc6550#section-6.7.4) and [RFC6551](https://tools.ietf.org/html/rfc6551))
```
Comment (miri64) [inline sys/include/net/gnrc/rpl.h:117]: Line-length

File: sys/include/net/gnrc/rpl.h [hunk 9]
```diff
@@ -82,6 +82,40 @@
...
+ *
+ *   e.g. for one alternative parent, set
+ *   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.mk}
+ *   CFLAGS += -DCONFIG_GNRC_IPV6_NIB_DEFAULT_ROUTER_NUMOF=2
+ *   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * TODO
+ * ------
+ *
+ * The GNRC RPL implementation only implements storing mode with OF0 ([RFC6552](https://tools.ietf.org/html/rfc6552)).
```
Comment (miri64) [inline sys/include/net/gnrc/rpl.h:110]: Line-length

File: sys/include/net/gnrc/rpl.h [hunk 10]
```diff
@@ -82,6 +82,40 @@
...
+ * The RPL routing header is parsed by the nodes, but anything else for non-storing mode is missing.
```
Comment (miri64) [inline sys/include/net/gnrc/rpl.h:111]: Just noticed this while reading this again. The RPL routing header is only parsed with the `gnrc_rpl_srh` module compiled in.

```suggestion
 * The RPL routing header is parsed by the nodes when the
 * [@c gnrc_rpl_srh](@ref net_gnrc_rpl_srh) module is used, but anything else
 * for non-storing mode is missing.
```

File: sys/include/net/gnrc/rpl.h (hunk 1 above)
Comment (miri64) [reply sys/include/net/gnrc/rpl.h:117]: Maybe add https://github.com/RIOT-OS/RIOT/pull/7231 or https://github.com/RIOT-OS/RIOT/pull/7231#issuecomment-651237343 as a reference here, so eager people don't start implementing when there is already a partly solution.

**Review by miri64 — COMMENTED** [review]
Please squash (including my last proposal)

**Review by miri64 — APPROVED** [review]
[All good](https://3876-7370241-gh.circle-artifacts.com/0/doc/group__net__gnrc__rpl.html) now

**Pull request merged**